### PR TITLE
NAS-103729 / 12 / Expose a new has_private attribute

### PR DESCRIPTION
**Pull request by sonicaj**
This commit introduces a new method on attribute 'has_private' which ensures that if we have a dict or dicts in a list's items, where any dict has an attribute marked as private, that is respected.

### CODE REVIEW

File: src/middlewared/middlewared/schema.py [hunk 1]
```diff
@@ -81,6 +81,9 @@ def clean(self, value):
                 raise Error(self.name, 'attribute required')
         return value
 
+    def has_private(self):
```
Comment (themylogin) [inline src/middlewared/middlewared/schema.py]: Can this be named `is_private` and just be used in `dump` like `if self.is_private(): return "***"`?

Comment (sonicaj) [reply src/middlewared/middlewared/schema.py]: So `is_private` as a name is quite similar to `private` which means the attribute in question is private. However, `has_private` serves a use case which indicates that the attribute has other children which are private and if the attribute is being used somewhere, the user of the attribute can decide how to handle the children case

Comment (themylogin) [reply src/middlewared/middlewared/schema.py]: If anything else is private, children are also private, how else can it be?

Comment (sonicaj) [reply src/middlewared/middlewared/schema.py]: How about attribute is not private but it has a few children which are, we do not want to hide the parent then, do we ?

Comment (themylogin) [reply src/middlewared/middlewared/schema.py]: When attribute is not private, we dump it, and children will hide themselves when dumping recursively. I've mean use it like this in `List.dump`

File: src/middlewared/middlewared/schema.py [hunk 2]
```diff
@@ -574,6 +577,9 @@ def __init__(self, name, *attrs, **kwargs):
                         raise ValueError(f"Attribute {attr.name} is not required and does not have default value, "
                                          f"this is forbidden in strict mode")
 
+    def has_private(self):
+        return self.private or any(i.private or i.has_private() for i in self.attrs.values())
```
Comment (themylogin) [inline src/middlewared/middlewared/schema.py]:
```suggestion
        return self.private or any(i.has_private() for i in self.attrs.values())
```

File: src/middlewared/middlewared/schema.py [hunk 3]
```diff
@@ -468,7 +471,7 @@ def clean(self, value):
         return value
 
     def dump(self, value):
-        if self.private or (self.items and any(item.private for item in self.items)):
+        if self.private or (self.items and any(item.private or item.has_private() for item in self.items)):
```
Comment (themylogin) [inline src/middlewared/middlewared/schema.py]:
```suggestion
        if self.has_private():
```

**Comment by themylogin** [comment]
```
def test__1():
    s = List('a', items=[List('b', items=[List('c', private=True)])])
    assert s.dump([[['d']]]) == '********'
```

File: src/middlewared/middlewared/schema.py (hunk 3 above)
Comment (themylogin) [reply src/middlewared/middlewared/schema.py]: Also you may promote this implementation to base class

File: src/middlewared/middlewared/schema.py [hunk 4]
```diff
@@ -467,11 +470,8 @@ def clean(self, value):
                     raise Error(self.name, 'Item#{0} is not valid per list types: {1}'.format(index, found))
         return value
 
-    def dump(self, value):
-        if self.private or (self.items and any(item.private for item in self.items)):
-            return "********"
-
-        return value
+    def has_private(self):
+        return self.private or (self.items and any(item.has_private() for item in self.items))
```
Comment (themylogin) [inline src/middlewared/middlewared/schema.py]: `any` will return `False` for empty list
```suggestion
        return self.private or any(item.has_private() for item in self.items)
```

File: src/middlewared/middlewared/pytest/unit/test_schema.py [hunk 5]
```diff
@@ -331,6 +331,14 @@ def dictargs(self, data):
         assert dictargs(self, value) == expected
 
 
+def test__schema_list_private():
+    for items, value in (
```
[the remark this answers is not in the document]
Comment (sonicaj) [reply src/middlewared/middlewared/pytest/unit/test_schema.py]: Thank you @themylogin for clarifying. I have addressed your concerns, please do let me know if something's amiss. Thanks again  :)

File: src/middlewared/middlewared/pytest/unit/test_schema.py [hunk 6]
```diff
@@ -331,6 +331,14 @@ def dictargs(self, data):
         assert dictargs(self, value) == expected
 
 
+@pytest.mark.parametrize('items,value', [
+    ([List('b', items=[List('c', private=True)])], [[['a']]]),
+    ([Dict('b', Str('c', private=True))], [{'c': 'secret'}])
+])
+def test__schema_list_private_items(items, value):
+    assert List('a', items=items).dump(value) == '********'
```
Comment (william-gr) [inline src/middlewared/middlewared/pytest/unit/test_schema.py]: Can we also add a test with a List with multiple Dict as items with private attributes?

Or not worth it?

Comment (william-gr) [reply src/middlewared/middlewared/pytest/unit/test_schema.py]: Err, nevermind me, i am dumb, that what is happening here : )

**Pull request merged**